ui/src/shared/components/MultiGrid/MultiGrid.tsx

```diff
@@ -0,0 +1,810 @@
+
+    const adjustedColumnIndex = Math.max(0, columnIndex - fixedColumnCount)
+    const adjustedRowIndex = Math.max(0, rowIndex - fixedRowCount)
+
+    if (this.bottomLeftGrid) {
+      this.bottomLeftGrid.recomputeGridSize({
+        columnIndex,
+        rowIndex: adjustedRowIndex,
+      })
+    }
+    if (this.bottomRightGrid) {
+      this.bottomRightGrid.recomputeGridSize({
+        columnIndex: adjustedColumnIndex,
+        rowIndex: adjustedRowIndex,
+      })
+    }
+
+    if (this.topLeftGrid) {
+      this.topLeftGrid.recomputeGridSize({
+        columnIndex,
+        rowIndex,
+      })
+    }
+
+    if (this.topRightGrid) {
+      this.topRightGrid.recomputeGridSize({
+        columnIndex: adjustedColumnIndex,
+        rowIndex,
+      })
+    }
+
+    this.leftGridWidth = null
+    this.topGridHeight = null
+    this.maybeCalculateCachedStyles(true)
+  }
+
+  public componentDidMount() {
+    const {scrollLeft, scrollTop} = this.props
+
+    if (scrollLeft > 0 || scrollTop > 0) {
+      const newState: Partial<State> = {}
+
+      if (scrollLeft > 0) {
+        newState.scrollLeft = scrollLeft
+      }
+
+      if (scrollTop > 0) {
+        newState.scrollTop = scrollTop
+      }
+
+      this.setState({...this.state, ...newState})
+    }
+    this.handleInvalidatedGridSize()
+  }
+
+  public componentDidUpdate() {
+    this.handleInvalidatedGridSize()
+  }
+
+  public render() {
+    const {
+      onScroll,
+      scrollLeft: scrollLeftProp, // eslint-disable-line no-unused-vars
+      onSectionRendered,
+      scrollToRow,
+      scrollToColumn,
+      scrollTop: scrollTopProp, // eslint-disable-line no-unused-vars
+      ...rest
+    } = this.props
+
+    this.prepareForRender()
+
+    // Don't render any of our Grids if there are no cells.
+    // This mirrors what Grid does,
+    // And prevents us from recording inaccurage measurements when used with CellMeasurer.
+    if (this.props.width === 0 || this.props.height === 0) {
+      return null
+    }
+
+    // scrollTop and scrollLeft props are explicitly filtered out and ignored
+
+    const {scrollLeft, scrollTop} = this.state
+
+    return (
+      <div style={this.containerOuterStyle}>
+        <div style={this.containerTopStyle}>
+          {this.renderTopLeftGrid(rest)}
+          {this.renderTopRightGrid({
+            ...rest,
+            ...onScroll,
+            scrollLeft,
+          })}
+        </div>
+        <div style={this.containerBottomStyle}>
+          {this.renderBottomLeftGrid({
+            ...rest,
+            onScroll,
+            scrollTop,
+          })}
+          {this.renderBottomRightGrid({
+            ...rest,
+            onScroll,
+            onSectionRendered,
+            scrollLeft,
+            scrollToColumn,
+            scrollToRow,
+            scrollTop,
+          })}
+        </div>
+      </div>
+    )
+  }
+
+  public cellRendererBottomLeftGrid = ({rowIndex, ...rest}) => {
```

Comment:
what do you think about adding input and output types to all of these non-React methods?

Reply:
👍